batch2/batch/driver/scheduler.py

```diff
@@ -19,6 +34,33 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def get_user_resources(self):
+        records = self.db.execute_and_fetchall(
+'''
+SELECT user, ready_cores_mcpu, running_cores_mcpu
+FROM user_resources
+WHERE ready_cores_mcpu > 0 OR running_cores_mcpu > 0;
+''')
+        return {record['user']: UserResources.from_record(record) for record in records}
+
+    async def compute_fair_share(self):
+        user_resources = await self.get_user_resources()
+        free_cores_mcpu = self.inst_pool.live_free_cores_mcpu
+        total_cores_mcpu = len(self.inst_pool.healthy_instances_by_free_cores) * self.inst_pool.worker_cores * 1000
+        ready_cores_mcpu = sum([resources.ready_cores_mcpu for resources in user_resources])
+
+        while free_cores_mcpu > 0 and ready_cores_mcpu > 0:
```

Comment:
I'm not sure why this isn't just a loop over users.  I think we would want to compute the shares in order of users from least cores to run to most cores to run, so cores for users that need fewer than a 1/N share can be allocated to users who need more cores.